src/data/Data.test.ts

```diff
@@ -190,21 +199,120 @@ describe('WorkspaceData', () => {
+    expect(screen.queryByText(/Preparing your data tables/)).toBeNull(); // no waiting message
+
+    // Act
+    await act(async () => {
+      jest.advanceTimersByTime(30000);
+    });
+
+    // Assert
+    expect(mockListAppsV2).toHaveBeenCalledTimes(2); // no further calls
+    expect(mockGetSchema).not.toHaveBeenCalled(); // never tried fetching schema, which depends on app status
+  });
+
+  it('stops polling for schema info if an error occurs while doing so', async () => {
+    // Arrange
+    const { workspaceDataProps, mockListAppsV2, mockGetSchema } = setup({
       workspace: defaultAzureWorkspace,
       status: 'RUNNING',
     });
 
+    mockGetSchema.mockRejectedValue(new Error('schema error'));
+
     // Act
     await act(async () => {
       render(h(WorkspaceData, workspaceDataProps));
     });
 
     // Assert
+    expect(mockListAppsV2).toHaveBeenCalledTimes(1); // only expected call, provisioning
+    expect(mockGetSchema).toHaveBeenCalledTimes(1); // only expected call, which resulted in an error
+    expect(screen.getByText(/An error occurred while preparing/)).toBeVisible(); // display error message
+
+    // Act
+    await act(async () => {
+      jest.advanceTimersByTime(30000);
+    });
+
+    // Assert
+    expect(mockListAppsV2).toHaveBeenCalledTimes(1); // no further invocations
+    expect(mockGetSchema).toHaveBeenCalledTimes(1); // no further invocations
+  });
+
+  it('polls for schema info until a PROVISIONING app is RUNNING', async () => {
+    // Arrange
+    const { workspaceDataProps, mockListAppsV2, listAppResponse, mockGetSchema } = setup({
+      workspace: defaultAzureWorkspace,
+      status: 'PROVISIONING',
+    });
+
+    mockListAppsV2
+      .mockResolvedValueOnce([{ ...listAppResponse, status: 'PROVISIONING' }])
+      .mockResolvedValueOnce([{ ...listAppResponse, status: 'PROVISIONING' }])
+      .mockResolvedValueOnce([{ ...listAppResponse, status: 'RUNNING' }]);
+
+    // Act
+    await act(async () => {
+      render(h(WorkspaceData, workspaceDataProps));
+    });
+
+    // Assert
+    expect(mockListAppsV2).toHaveBeenCalledTimes(1); // initial call, provisioning
+    expect(mockGetSchema).not.toHaveBeenCalled(); // don't fetch schema yet
+    expect(screen.queryByText(/Select a data type/)).toBeNull();
+    expect(screen.getByText(/Preparing your data tables/)).toBeVisible();
+
+    // Act
+    await act(async () => {
+      jest.advanceTimersByTime(30000);
+    });
+
+    // Assert
+    expect(mockListAppsV2).toHaveBeenCalledTimes(2); // second call, still provisioning
+    expect(mockGetSchema).not.toHaveBeenCalled(); // don't fetch schema yet
+    expect(screen.queryByText(/Select a data type/)).toBeNull();
+    expect(screen.getByText(/Preparing your data tables/)).toBeVisible();
+
+    // Act
+    await act(async () => {
+      jest.advanceTimersByTime(30000);
+    });
+
+    // Assert
```

Comment:
Does AAAAAAA violate the spirit of AAA if it's testing polling? :D

Reply:
_I_ think it's fine. I prefer to write tests this way vs storing intermediate results in order to have one act and one assert section.